### ISSUE TITLE
Tier modification doesn't work

### ISSUE THREAD

**Issue by ChiefArug** [KubeJS-Mods/KubeJS#662]
### Minecraft Version

1.19.2

### KubeJS Version

56751b429ee50ab05cf6a308f4646a67f5c3e1bb

### Rhino Version

rhino-forge-1902.2.2-build.268.jar

### Architectury Version

6.5.82

### Forge/Fabric Version

43.2.11

### Describe your issue

This script doesn't work. 
The first one is visible in tooltip and when testing against mobs. Confirmed the second one through a debugger, `stack.getItem().getEnchantmentValue()` returns 14, the default.

However, for both of them, they cannot be repaired at all. Even the wooden sword. Removing the tier part of the modification fixes that.
```js
ItemEvents.modification(event => {
  event.modify('golden_sword', item => {
    item.tier = tier => {
        tier.speed = 12
        tier.attackDamageBonus = 10
        tier.repairIngredient = '#forge:storage_blocks/gold'
        tier.level = 3
    }
  })
  event.modify('wooden_sword', item => {
    item.tier = tier => {
        tier.enchantmentValue = 30
    }
  })
})
```

**Comment by SiverDX** [comment]
> Removing the tier part of the modification fixes that.

what does that mean? how would i have to change the script to make it work?
`item.tier.tierOptions.repairIngredient` or `item.tier.repairIngredient` or `item.repairIngredient` (i.e. not doing `= tier =>`) do not work and crash the game

``` js 
item.setTier(tierOptions => {
    tierOptions.repairIngredient = ...
}
```
also doesn't work

`item.tier.setRepairIngredient` does not exist

**Comment by ChiefArug** [comment]
> what does that mean?

The thing I was talking about with that line was that trying to modify the tier of anything resulted in the repair ingredient of that tier not working anymore, even if I did not try to modify the repair ingredient.

> how would i have to change the script to make it work?

You can't. The entire point of this issue is that it doesn't work.

> `item.tier.setRepairIngredient` does not exist

It only exists on KubeJS' `MutableToolTier` class, it doesn't exist by default on other Tier implementations.

**Comment by SiverDX** [comment]
But what does 

`Kinda broken at the moment! https://github.com/KubeJS-Mods/KubeJS/issues/662. Use the non tier methods instead.`

mean then? Does it just mean "you cannot modify anything regarding tiers at the moment" or does it mean "you can use some? method of the item itself to modify tiers"?

**Comment by ChiefArug** [comment]
It means you cannot modify tiers at the moment.

**Comment by southern-ttv** [comment]
Anything going on with this been about a year.

**Comment by ChiefArug** [comment]
No progress that I know of.

**Comment by velox-resurrected** [comment]
Added a note for others that end up on this bug report page, this is also broken for 1.20.1 as well.

**Comment by Farcr** [comment]
How am i supposed to change an item's durability and other tier properties if i can't modify tiers?

**Comment by miner-sd** [comment]
is this ever getting fixed? it seems pretty important...

**Comment by Disketaa** [comment]
bump

**Comment by NatSquared** [comment]
@ChiefArug are you able to use iron to repair the custom tool? I have a hunch that this is due to all tiers being based on `net.minecraft.world.item.Tiers.IRON` by default, which might be immutable or something
if you don't have time to test I'll do it sometime in the next few days

**Comment by Auseawesome** [comment]
> @ChiefArug are you able to use iron to repair the custom tool? I have a hunch that this is due to all tiers being based on `net.minecraft.world.item.Tiers.IRON` by default, which might be immutable or something
> if you don't have time to test I'll do it sometime in the next few days
> Edit: sorry, commented hastily. I could be wrong here - the Iron default might only apply to _new_ custom tiers, not modifying existing ones

This issue could potentially be related to the immutability of tiers though, that could be why tier modification is failing entirely...

**Comment by pietro-lopes** [comment]
Are you guys still talking about 1.19.2?
Asking this because iirc tiers are different on 1.20.1 also in 1.21.1

**Comment by MaxNeedsSnacks** [comment]
> Are you guys still talking about 1.19.2? Asking this because iirc tiers are different on 1.20.1 also in 1.21.1

Yes, which is precisely why I was able to fix this in like 5 minutes on 1.21 and have gone crazy thinking about how to do this cross platform in 1.19/1.20…

**Comment by Tincyclopedia** [comment]
Also occurs in 1.21...

**Comment by pietro-lopes** [comment]
> Also occurs in 1.21...

Create an issue and examples of what you are trying to do in 1.21.1, I'm pretty sure it works fine on 1.21.1 iirc

**Comment by lifekill** [comment]
Is there an alternative workaround for this issue? I'm still learning KubeJS/Java, and this is making me look at Visual Studio feel like hell

**Comment by pietro-lopes** [comment]
Join discord and check for examples

**Comment by dninemfive** [comment]
> > Also occurs in 1.21...
> 
> Create an issue and examples of what you are trying to do in 1.21.1, I'm pretty sure it works fine on 1.21.1 iirc

See [here](https://github.com/libcraft-games/LibPack/blob/main/kubejs/startup_scripts/item/modification/netherite_scrap_repair.js) for an example of a use which appears to be nonfunctional in-game. This script successfully runs but no changes are applied to the items in question, their repair ingredient is still netherite ingots.

Maybe things have changed from the documentation, but in that case the documentation needs to be updated.

> Join discord and check for examples

It seems like a shame to restrict information to the discord when you have a pretty good (if incomplete) documentation site. Last time i was active in modding i hit the server cap on discord just from joining servers for stuff like this.

**Comment by DragonacyCX** [comment]
Yo everybody there is a workaround mod for changing pickaxe tiers, Better Harvest Levels by TysonTheEmber, tested it and it works well. https://www.curseforge.com/minecraft/mc-mods/better-harvest-levels

**Comment by OctavoPE** [comment]
Nothing yet still? How's this possible?

**Comment by mIT-becker** [comment]
Greetings together,
there are two walkarounds (for 1.21.1) to set the repair item and the mining level.

You can add a script in the startup_scripts/ and configure it:
```
// Sets RepairIngredient:
ItemEvents.modification(event => {
  event.modify('<namespace>:<itemID>', item => {
    item.tier = tier => {
      tier.repairIngredient = '<namespace>:<itemID>' //namespace f.e. kubejs and itemID for example copper_pickaxe.
    };
  });
});

// Sets Mining Level:
ItemEvents.modification(event => {
	const modifyTier = (itemID, tagString) => {
		event.modify(itemID, item => {
			item.setTier(tier => {
				tier.setIncorrectBlocksForDropsTag(tagString);
			});
		});
	};
	modifyTier('<namespace>:<itemID>', 'minecraft:incorrect_for_<TIER>_tool'); //change TIER to wished tier f.e. diamond.
});
```

For Help just ask, I hoped this helps you. Have a nice day!
